util/EIP-712/src/encode.rs

```diff
@@ -0,0 +1,414 @@
+				}
+				types.insert(&*field.type_);
+			}
+		}
+	};
+
+	return Some(deps)
+}
+
+fn encode_type(message_type: &str, message_types: &MessageTypes) -> Result<String> {
+	let deps = {
+		let mut temp = build_dependencies(message_type, message_types).ok_or_else(|| ErrorKind::NonExistentType)?;
+		temp.remove(message_type);
+		let mut temp = temp.into_iter().collect::<Vec<_>>();
+		(&mut temp[..]).sort_unstable();
+		temp.insert(0, message_type);
+		temp
+	};
+
+	let encoded = deps
+		.into_iter()
+		.filter_map(|dep| {
+			message_types.get(dep).map(|field_types| {
+				let types = field_types
+					.iter()
+					.map(|value| format!("{} {}", value.type_, value.name))
+					.join(",");
+				return format!("{}({})", dep, types);
+			})
+		})
+		.collect::<Vec<_>>()
+		.concat();
+	Ok(encoded)
+}
+
+fn type_hash(message_type: &str, typed_data: &MessageTypes) -> Result<H256> {
+	Ok(keccak(encode_type(message_type, typed_data)?))
+}
+
+fn encode_data(
+	parser: &Parser,
+	message_type: &Type,
+	message_types: &MessageTypes,
+	value: &Value,
+	field_name: Option<&str>
+) -> Result<Vec<u8>>
+{
+	let encoded = match message_type {
+		Type::Array {
+			inner,
+			length
+		} => {
+			let mut items = vec![];
+			let values = value.as_array().ok_or_else(|| serde_error("array", field_name))?;
+
+			// check if the type definition actually matches
+			// the length of items to be encoded
+			if *length != 0 && values.len() as u64 != *length {
+				let array_type = format!("{}[{}]", *inner, length);
+				return Err(ErrorKind::UnequalArrayItems(*length, array_type, values.len() as u64))?
+			}
+
+			for item in values {
+				let mut encoded = encode_data(parser, &*inner, &message_types, item, field_name)?;
+				items.append(&mut encoded);
+			}
+
+			keccak(items).to_vec()
+		}
+
+		Type::Custom(ref ident) if message_types.get(&*ident).is_some() => {
+			let type_hash = (&type_hash(ident, &message_types)?).to_vec();
+			let mut tokens = encode(&[EthAbiToken::FixedBytes(type_hash)]);
+
+			for field in message_types.get(ident).expect("Already checked in match guard; qed") {
+				let value = &value[&field.name];
+				let type_ = parser.parse_type(&*field.type_)?;
+				let mut encoded = encode_data(parser, &type_, &message_types, &value, Some(&*field.name))?;
+				tokens.append(&mut encoded);
+			}
+
+			keccak(tokens).to_vec()
+		}
+
+		Type::Bytes => {
+			let string = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if string.len() < 2 {
+				return Err(ErrorKind::HexParseError(
+					format!("Expected a 0x-prefixed string of even length, found {} length string", string.len()))
+				)?
+			}
+			let bytes = (&string[2..])
+				.from_hex::<Vec<u8>>()
+				.map_err(|err| ErrorKind::HexParseError(format!("{}", err)))?;
+			let bytes = keccak(&bytes).to_vec();
+
+			encode(&[EthAbiToken::FixedBytes(bytes)])
+		}
+
+		Type::Byte(_) => {
+			let string = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if string.len() < 2 {
```

Comment:
do we want to check if the prefix is actually `"0x"`? and maybe extract this check into a function